src/StreamUtils.bsv

```diff
@@ -0,0 +1,367 @@
+import Vector::*;
+import FIFOF::*;
+import GetPut::*;
+import SemiFifo::*;
+
+typedef 8 BYTE_WIDTH;
+typedef TLog#(BYTE_WIDTH) BYTE_WIDTH_WIDTH;
+typedef TMul#(4, BYTE_WIDTH) DWORD_WIDTH;
+
+typedef 2 CONCAT_STREAM_NUM;
+
+typedef 512 DATA_WIDTH;
+typedef TDiv#(DATA_WIDTH, BYTE_WIDTH) BYTE_EN_WIDTH;
+typedef 'hFFFFFFFFFFFFFFFF MAX_BYTE_EN;
+
+typedef Bit#(DATA_WIDTH) Data;
+typedef Bit#(BYTE_EN_WIDTH) ByteEn;
+typedef Bit#(TAdd#(1, TLog#(DATA_WIDTH))) DataBitPtr;
+typedef Bit#(TAdd#(1, TLog#(BYTE_EN_WIDTH))) DataBytePtr;
+
+typedef 32 STREAM_SIZE_WIDTH;
+typedef UInt#(STREAM_SIZE_WIDTH) StreamSize;
+typedef Bit#(TAdd#(1, TLog#(STREAM_SIZE_WIDTH))) StreamSizeBitPtr;
+
+typedef struct {
+    Data data;
+    ByteEn byteEn;
+    Bool isFirst;
+    Bool isLast;
+} DataStream deriving(Bits, Bounded, Eq, FShow);
+
+typedef struct {
+    DataStream stream;
+    DataBytePtr bytePtr;
+} StreamWithPtr deriving(Bits, Bounded, Eq, FShow);
+
+interface StreamConcat;
+    interface FifoIn#(DataStream)  inputStreamFirstFifoIn;
+    interface FifoIn#(DataStream)  inputStreamSecondFifoIn;
+    interface FifoOut#(DataStream) outputStreamFifoOut;
+endinterface
+
+interface StreamSplit;
+    interface FifoIn#(DataStream)  inputStreamFifoIn;
+    interface FifoIn#(StreamSize)  splitLocationFifoIn;
+    interface FifoOut#(DataStream) outputStreamFifoOut;
+endinterface
+
+function DataBytePtr convertByteEn2BytePtr (ByteEn byteEn);
+    ByteEn byteEnTemp = byteEn;
+    DataBytePtr ptr = 0;
+    while (byteEnTemp > 0) begin
+        byteEnTemp = byteEnTemp >> 1;
+        ptr = ptr + 1;
+    end
+    return ptr;
+endfunction
+
+function DataStream getEmptyStream ();
+    return DataStream{
+        data: 0,
+        byteEn: 0,
+        isFirst: False,
+        isLast: True
+    };
+endfunction
+
+function DataBitPtr getMaxBitPtr ();
+    return fromInteger(valueOf(DATA_WIDTH));
+endfunction
+
+function DataBytePtr getMaxBytePtr ();
+    return fromInteger(valueOf(BYTE_EN_WIDTH));
+endfunction
+
+// Concat two DataStream frames into one. StreamA.isLast must be True, otherwise the function will return a empty frame to end the stream.
+function Tuple3#(DataStream, DataStream, DataBytePtr) getConcatStream (DataStream streamA, DataStream streamB, DataBytePtr bytePtrA, DataBytePtr bytePtrB);
```

Comment:
please do defensive coding. Add an assert here since you have requirements for input params.
It will save you from endless debugging.

see 

https://github.com/datenlord/blue-rdma/blob/ffdcc150ae50cd6c94aee6f683d5f2133a0c662e/src/PrimUtils.bsv#L148

and

https://github.com/datenlord/blue-rdma/blob/ffdcc150ae50cd6c94aee6f683d5f2133a0c662e/src/PayloadConAndGen.bsv#L999